src/PostgREST/Request/QueryParams.hs

```diff
@@ -0,0 +1,489 @@
+    , qsLogic          :: [(EmbedPath, LogicTree)]
+    -- ^ &and and &or parameters used for complex boolean logic
+    , qsColumns        :: Maybe (S.Set FieldName)
+    -- ^ &columns parameter and payload
+    , qsSelect         :: [Tree SelectItem]
+    -- ^ &select parameter used to shape the response
+    , qsFilters        :: [(EmbedPath, Filter)]
+    -- ^ Filters on the result from e.g. &id=e.10
+    , qsFiltersRoot    :: [(EmbedPath, Filter)]
+    -- ^ Subset of the filters that apply on the root table
+    , qsFiltersNotRoot :: [(EmbedPath, Filter)]
+    -- ^ Subset of the filters that do not apply on the root table
+    , qsFilterFields   :: S.Set FieldName
+    -- ^ Set of fields that filters apply to
+    , qsOnConflict     :: Maybe [FieldName]
+    -- ^ &on_conflict parameter used to upsert on specific unique keys
+    }
+
+-- |
+-- Parse query parameters from a query string like "id=eq.1&select=name".
+--
+-- The canonical representation of the query string has paramters sorted alphabetically:
+--
+-- >>> qsCanonical <$> parse "a=1&c=3&b=2&d"
+-- Right "a=1&b=2&c=3&d="
+--
+-- 'select' is a reserved parameter that selects the fields to be returned:
+--
+-- >>> qsSelect <$> parse "select=name,location"
+-- Right [Node {rootLabel = (("name",[]),Nothing,Nothing,Nothing,Nothing), subForest = []},Node {rootLabel = (("location",[]),Nothing,Nothing,Nothing,Nothing), subForest = []}]
+--
+-- Filters are parameters whose value contains an operator, separated by a '.' from its value:
+--
+-- >>> qsFilters <$> parse "a.b=eq.0"
+-- Right [(["a"],Filter {field = ("b",[]), opExpr = OpExpr False (Op OpEqual "0")})]
+--
+-- If the operator specified in a filter does not exist, parsing the query string fails:
+--
+-- >>> qsFilters <$> parse "a.b=noop.0"
+-- Left (QueryParamError "\"failed to parse filter (noop.0)\" (line 1, column 6)" "unknown single value operator noop")
+parse :: ByteString -> Either QueryParamError QueryParams
+parse qs =
+  QueryParams
+    canonical
+    params
+    ranges
+    <$> pRequestOrder `traverse` order
+    <*> pRequestLogicTree `traverse` logic
+    <*> pRequestColumns columns
+    <*> pRequestSelect select
+    <*> pRequestFilter `traverse` filters
+    <*> pRequestFilter `traverse` filtersRoot
+    <*> pRequestFilter `traverse` filtersNotRoot
+    <*> pure (S.fromList (fst <$> filters))
+    <*> sequenceA (pRequestOnConflict <$> onConflict)
+  where
+    logic = filter (endingIn ["and", "or"] . fst) nonemptyParams
+    select = fromMaybe "*" $ lookupParam "select"
+    onConflict = lookupParam "on_conflict"
+    columns = lookupParam "columns"
+    order = filter (endingIn ["order"] . fst) nonemptyParams
+    -- Replace .limit and .offset endings to be able to match those params later in a map
+    limits = first (replaceLast "") <$> filter (endingIn ["limit"] . fst) nonemptyParams
+    offsets = first (replaceLast "") <$> filter (endingIn ["offset"] . fst) nonemptyParams
```

Comment:
```suggestion
    limits = first (replaceLast "range") <$> filter (endingIn ["limit"] . fst) nonemptyParams
    offsets = first (replaceLast "range") <$> filter (endingIn ["offset"] . fst) nonemptyParams
```

Seems like CI failed with the empty string - maybe like this, instead?

Reply:
Hm, there seems to be some particular dependency on this being renamed `limit`, tests fail otherwise. I'll add a comment and we'll investigate in a separate PR